### ISSUE TITLE
Apparent bug with certain comments

### ISSUE THREAD

**Issue by dkon1** [yihui/knitr#1855]
When I knit this document, the following chunks generate very strange output, not consistent with the output generated by running the chunks within the Rmd document. 

In the chunk below, the echo shows repeated lines of code, and the output of the print(hey) command is 1 rather than 2.
````
```{r}
hey <- 1
hey <- 2
#line 1
print(hey)
```
````

If I add a comment at the top of the code chunk, the knitted file still includes duplicated lines of code, but the output of print(hey) is correct.

````
```{r}
# any comment here
hey <- 1
hey <- 2
#line 1
print(hey)
```
````

In the chunk below, the echo shows a whole bunch of NAs within the code and the print(hey) output is correct.

````
```{r}
hey <- 1
hey <- 2
#line 123
print(hey)
```
````

This output is the same whether I knit to html or pdf. This strange behavior disappears if I change the comment to anything except #line... - even putting in a space after # makes it go away.

---

By filing an issue to this repo, I promise that

- [X] I have fully read the issue guide at https://yihui.org/issue/.
- [X] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('knitr')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('yihui/knitr')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [X] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

**Comment by dkon1** [comment]
> xfun::session_info('knitr')
R version 4.0.0 (2020-04-24)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Catalina 10.15.4

Locale: en_US.UTF-8 / en_US.UTF-8 / en_US.UTF-8 / C / en_US.UTF-8 / en_US.UTF-8

Package version:
  evaluate_0.14   glue_1.4.1      graphics_4.0.0  grDevices_4.0.0
  highr_0.8       knitr_1.28      magrittr_1.5    markdown_1.1   
  methods_4.0.0   mime_0.9        stats_4.0.0     stringi_1.4.6  
  stringr_1.4.0   tools_4.0.0     utils_4.0.0     xfun_0.14      
  yaml_2.2.1

**Comment by shrektan** [comment]
I can reproduce this...

**Comment by cderv** [comment]
This is a weird behavior for sure. But this comes from below knitr, from the `evaluate` 📦 

Here is the reproducible example from the "weird" parsing behavior, and so the "weird" evaluation.

``` r
chunk1 <- c("hey <- 1", "hey <- 2", "#line 1", "print(hey)")
chunk2 <- c("# any comment here", "hey <- 1", "hey <- 2", "#line 1", "print(hey)")
chunk3 <- c("hey <- 1", "hey <- 2", "# line 123", "print(hey)")

# There is something off in the parsing
# for those 2, and hence its evaluation
evaluate::parse_all(chunk1)
#>          src       expr
#> 1 hey <- 1\n   hey <- 1
#> 2 hey <- 1\n print(hey)
#> 3 hey <- 2\n           
#> 4 hey <- 2\n   hey <- 2
#> 5  #line 1\n           
#> 6 print(hey)
evaluate::parse_all(chunk2)
#>                    src       expr
#> 1 # any comment here\n           
#> 2 # any comment here\n print(hey)
#> 3           hey <- 1\n           
#> 4           hey <- 1\n   hey <- 1
#> 5           hey <- 2\n           
#> 6           hey <- 2\n   hey <- 2
#> 7            #line 1\n           
#> 8           print(hey)

# expected parsing 
evaluate::parse_all(chunk3)
#>            src       expr
#> 1   hey <- 1\n   hey <- 1
#> 2   hey <- 2\n   hey <- 2
#> 3 # line 123\n           
#> 4   print(hey) print(hey)
```

<sup>Created on 2020-06-17 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

So I need to dig further and understand what happens, then it will be something to report in `evaluate` I think... 

Thanks.

_Note: @dkon1, I took the liberty to edit your question to get the proper formating. You need to add the four backtick before and after each chunk you want as-is in md format: https://yihui.org/issue/#please-format-your-issue-correctly_

**Comment by cderv** [comment]
Ok it happens that `#line` is a specific directive in base R 
Look at `?srcfile`

So having a `#line` will have a specific behavior in the R code parsing then associated the source and parsed expression. As evaluate associate src and parsed expression, that knitr uses, it causes the "weird" behavior you see. 

Here an example to show how R see a line beginning by `#line <something>`
``` r
getParseData(parse(text = c("#line 1", "#line", "# line 1")))
#>   line1 col1 line2 col2 id parent          token terminal     text
#> 1     1    1     1    7  1      0 LINE_DIRECTIVE     TRUE  #line 1
#> 4     2    1     2    5  4      0        COMMENT     TRUE    #line
#> 7     3    1     3    8  7      0        COMMENT     TRUE # line 1
```

<sup>Created on 2020-06-17 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

Look at how `#line 1` is seen by R as a **LINE_DIRECTIVE** token, and not a **COMMENT**.
Adding a space or just `#line` will make it seen as a comment. 

This is not a knitr issue here. Just that `#line <something>` in R code as a special meaning... Not clear to me yet though. 🤔 

So I would say, don't forget the space after the `#` 😄 

Thanks for the question, I learn something today !

**Comment by dkon1** [comment]
Thank you @cderv that is very interesting and I had no idea that R could interpret anything after `#`!

But I'm still unclear why the same code does not cause any problems when running within the RStudio environment (by clicking the green arrow on the code chunk) but does cause problems when knitting. If it's a base R directive, shouldn't it be interpreted the same?

**Comment by cderv** [comment]
> But I'm still unclear why the same code does not cause any problems when running within the RStudio environment (by clicking the green arrow on the code chunk) but does cause problems when knitting. If it's a base R directive, shouldn't it be interpreted the same

I think the RStudio IDE is parsing differently to execute the chunk using the "Run Current Chunk" button. Maybe not with `evaluate`. If you try to render with the Knit button, you also get the error.

It would be possible to bypass this `#line ` special R parsing in `evaluate` 📦 maybe... 🤔 
Don't know if it is useful to have it or not...

**Comment by yihui** [comment]
I don't know what `LINE_DIRECTIVE` really means (this is the first time I have heard of it), but if we are sure it is a comment, we can probably treat it as comment in the **evaluate** package.

**Comment by cderv** [comment]
I also discovered it while investigating this issue. 
Here is the most clearer meaning I found about it in R journal about **Source References**
https://journal.r-project.org/archive/2010-2/RJournal_2010-2_Murdoch.pdf

> The #line directive
In some cases, R source code is written by a program, not by a human being. For example, Sweave() extracts lines of code from Sweave documents before sending the lines to R for parsing and evaluation. To support such preprocessors, the R 2.10.0 parser recognizes a new directive of the form #line nn "filename" where nn is an integer. As with the same-named directive in the C language, this tells the parser to assume that the next line of source is line nn from the given filename for the purpose of constructing source references. The Sweave() function doesn’t currently make use of this, but in the future, it (and other preprocessors) could output #line directives so that source references and syntax errors refer to the original source location rather than to an intermediate file.
The #line directive was a late addition to R 2.10.0. Support for this in Sweave() appeared in R 2.12.0.

This comes from the C language where you can find from reference to it. ([Example](
https://docs.microsoft.com/en-us/cpp/preprocessor/hash-line-directive-c-cpp?view=vs-2019))

This is an interesting feature (but very advanced!) that allow to reference the original source code, even if it happens in from intermediary file. (Seems to be used in Sweave)

I am not sure this is useful or not. At least, I think we can consider that codes evaluated from a chunk won't ever use this line directive feature. I don't why one woud use that in a chunk.

**Comment by dkon1** [comment]
Thanks for the explanation Christophe! To be clear, I found this strange feature because a student of mine (in a basic course) used something like `#line 1` as a comment in their assignment and was getting strange errors they couldn't parse, so I took a look and found that it was caused by this comment. From a basic user's perspective I think it would be good to remove this functionality from `evaluate`.
